src/lib/autocomplete/autocomplete-trigger.ts

```diff
@@ -431,7 +431,9 @@ export class MatAutocompleteTrigger implements ControlValueAccessor, OnDestroy {
   private _subscribeToClosingActions(): Subscription {
     const firstStable = this._zone.onStable.asObservable().pipe(take(1));
     const optionChanges = this.autocomplete.options.changes.pipe(
-      tap(() => this._positionStrategy.reapplyLastPosition()),
+      // Adding/removing options may mean the overlay now exceeds the viewport,
+      // so update its position
+      tap(() => this._overlayRef!.updatePosition()),
```

Comment:
The old behavior was deliberate. We decided not to flip the position further, because it can be distracting for the user. E.g. if you type something, you get the dropdown on the bottom, then you type another character and the dropdown could flip to the top, then you type again and it could end up on the bottom again.

Reply:
Ah, I see. In my use case the behaviour is indesirable because the panel is _always_ outside the viewport (since the results are loaded dynamically after the user focuses the input). Maybe it could be an option?

Reply:
In general we try to avoid adding APIs for such specific use cases, because they tend to pile up and make things hard to refactor. Would making the panel flexible (having it become scrollable once it hits the viewport edge) be a potential solution?

Reply:
It would partially help, but still wouldn't completely fix it in some cases (e.g. when the panel height is less than the height of one option).

To be honest I don't think it's a hugely unusual use case. You can see in [this demo](https://stackblitz.com/edit/angular-material2-issue-7v48o6?file=app/app.component.ts) that it's trivial to reproduce - all it requires is that the autocomplete is near the bottom of the viewport, and that the options are loaded dynamically when the user types. 

Specifically it is a problem when the options are "additive", in that options are only populated once the user starts typing in the autocomplete, rather than the alternate use case of the user's input filtering a fixed list of options.

There are examples of this use case all over the web, one prominent example being the main search bar on google.com - if you were to implement this using `MatAutocomplete` with the component placed further down the viewport, it is likely that in the current implementation the UX would be broken.

Reply:
I see what you mean, but I'm not sure that either of the solutions (always updating the position or having an opt-out option) would be optimal. We may be able to do something more intuitive where we don't determine the position until the first time a set of options comes in.

Reply:
I did consider this, but it still wouldn't really work for typeaheads where the number of results can increase past the initial count. This is actually true for my use case, where the user is tagging an entity and the list of options are (list of existing tags that match the input value) + "Create new tag: [input value]". When the user starts typing, they immediately have the "Create new tag" option, but after some time the API call completes and the relevant existing tags are added to the list of options.

When you say opt-out, do you mean that the current behaviour would be default? Because that's what I was imagining, maybe a `matAutocompleteRepositionPanel` option that is false by default.

Reply:
I do have a similar use case : my options are added dynamically. If none is in cache when the panel initially opens, it contains a "search in progress" message. This is to inform the user that this field will have assistance soon.
If the field is in the bottom of the page, it is almost always outside of the viewport since it opens with a small initial panel.
You could maybe expose a reposition() method in the matAutocomplete API so that we could trigger it programmatically (instead of a configuration option). 
Better still, the position strategy could choose to change position only if the content doesn't fit in the current one : in your example were the options change back and forth with the user input it would only jump once.

I understand the need for the smallest possible API surface and the consideration of the majority where several consecutive jumps would ruin the user experience, but a piece of UI that is inaccessible because it is outside of the viewport in my opinion (you wouldn't write in the docs that an autocomplete cannot be placed less than its maximum height from the bottom of the viewport when its content changes dynamically).

Reply:
@crisbeto How would you feel about a PR which instead added a `reposition()` method as described above?

Reply:
I'd be ok with exposing a method for triggering the reposition yourself. I'd call it `updatePosition` so it aligns with the `OverlayRef`.